<p><html><body><a href="http://asmeurersympy.wordpress.com/2009/07/20/modifying-a-list-while-looping-through-it-in-python/">Every once in a while</a>, seemingly really simple Python code does something completely unexpected for me. Look at the following snippet of Python code.  This is run straight from the 2.6.5 interpreter, with no other commands executed.  Do you notice anything strange?</p>
<p>[code language="py"]</p>
<p>$python</p>
<p>Python 2.6.5 (r265:79359, Mar 24 2010, 01:32:55) </p>
<p>[GCC 4.0.1 (Apple Inc. build 5493)] on darwin</p>
<p>Type "help", "copyright", "credits" or "license" for more information.</p>
<p>&gt;&gt;&gt;; l = lambda i: a[i]</p>
<p>&gt;&gt;&gt; l</p>
<p>&lt;function at="" 0x39e7f0=""&gt;</p>
<p>&gt;&gt;&gt; H = [(1, 2), (3, 4)]</p>
<p>&gt;&gt;&gt; [l(0) + l(1) for a in H]</p>
<p>[3, 7]</p>
<p>[/code]</p>
<p>Did you spot it?  Here is a hint. Running a different but similar session:</p>
<p>[code language="py"]</p>
<p>$python</p>
<p>Python 2.6.5 (r265:79359, Mar 24 2010, 01:32:55) </p>
<p>[GCC 4.0.1 (Apple Inc. build 5493)] on darwin</p>
<p>Type "help", "copyright", "credits" or "license" for more information.</p>
<p>&gt;&gt;&gt; l = lambda i: a[i]</p>
<p>&gt;&gt;&gt; l</p>
<p>&lt;function at="" 0x39e7f0=""&gt;</p>
<p>&gt;&gt;&gt; l(0)</p>
<p>Traceback (most recent call last):
  File "", line 1, in 
  File "", line 1, in 
NameError: global name 'a' is not defined</p>
<p>[/code]</p>
<p>Do you see it now?  I defined the lambda function <code>l</code> in terms of <code>a</code> without defining first defining <code>a</code>!  And furthermore, it just works when <code>a</code> is defined.  This is actually independent of the fact that we are working in a list comprehension, as this continuation of the previous session shows:</p>
<p>[code language="py"]</p>
<p>&gt;&gt;&gt; a = [3, 4, 5]</p>
<p>&gt;&gt;&gt; l(0)</p>
<p>3</p>
<p>[/code]</p>
<p>But I want to expand on the list comprehension example, because there even more bizzare things going on here.  Restarting a new session again:</p>
<p>[code language="py"]</p>
<p>$python</p>
<p>Python 2.6.5 (r265:79359, Mar 24 2010, 01:32:55) </p>
<p>[GCC 4.0.1 (Apple Inc. build 5493)] on darwin</p>
<p>Type "help", "copyright", "credits" or "license" for more information.</p>
<p>&gt;&gt;&gt; l = lambda i: a[i]</p>
<p>&gt;&gt;&gt; H = [(1, 2), (3, 4)]</p>
<p>&gt;&gt;&gt; [l(0) + l(1) for a in H]</p>
<p>[3, 7]</p>
<p>&gt;&gt;&gt; (l(0) + l(1) for a in H)</p>
<p>&lt;generator object="" at="" 0x3a4350=""&gt;</p>
<p>&gt;&gt;&gt; list((l(0) + l(1) for a in H))</p>
<p>[7, 7]</p>
<p>[/code]</p>
<p>So, if you are astute and have been using Python for long enough, you should be able to catch what is going on here.  If you don't know, here is a hint (continuation of previous session):</p>
<p>[code language="py"]</p>
<p>&gt;&gt;&gt; a</p>
<p>(3, 4)</p>
<p>[/code]</p>
<p>So, as you may know, in Python 2.6 and earlier, list comprehension index variables "leek" into the local namespace.  The strange thing here is that although the list comprehension would reset it, the generator version does not.  Well, normally, it does do this:</p>
<p>[code language="py"]</p>
<p>&gt;&gt;&gt; x = 1</p>
<p>&gt;&gt;&gt; [x for x in range(10)]</p>
<p>[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]</p>
<p>&gt;&gt;&gt; x</p>
<p>9</p>
<p>&gt;&gt;&gt; del x</p>
<p>&gt;&gt;&gt; list((x for x in range(10)))</p>
<p>[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]</p>
<p>&gt;&gt;&gt; x</p>
<p>Traceback (most recent call last):
  File "", line 1, in 
NameError: name 'x' is not defined</p>
<p>&gt;&gt;&gt; x = 1</p>
<p>&gt;&gt;&gt; list((x for x in range(10)))</p>
<p>[0, 1, 2, 3, 4, 5, 6, 7, 8, 9]</p>
<p>&gt;&gt;&gt; x</p>
<p>1</p>
<p>[/code]</p>
<p>So the above bit has something to do with the way the <code>lambda</code> function was defined with the <code>a</code>.  By the way, here is what happens with the generator comprehension (is that what these are called?) if <code>a</code> is not defined:</p>
<p>[code language="py"]</p>
<p>&gt;&gt;&gt; del a</p>
<p>&gt;&gt;&gt; list((l(0) + l(1) for a in H))</p>
<p>Traceback (most recent call last):
  File "", line 1, in 
  File "", line 1, in 
  File "", line 1, in 
NameError: global name 'a' is not defined</p>
<p>[/code]</p>
<p>This is how I discovered this.  I had defined a lambda function using an variable that was then passed to a list comprehension that used this variable as the index without realizing it. But then I tried converting this into a generator comprehension to see if it would be faster, and got the above error.  </p>
<p>Finally, since the "feature" of leaking list comprehension loop variables into the local namespace is <a href="http://docs.python.org/release/3.0.1/whatsnew/3.0.html#changed-syntax">going away</a> in Python 3, I expected things to behave at least a little differently in Python 3.  I tried the above in a Python 3.1.2 interpreter and got the following:</p>
<p>[code language="py"]</p>
<p>$python3</p>
<p>Python 3.1.2 (r312:79147, Mar 23 2010, 22:02:05) </p>
<p>[GCC 4.2.1 (Apple Inc. build 5646) (dot 1)] on darwin</p>
<p>Type "help", "copyright", "credits" or "license" for more information.</p>
<p>&gt;&gt;&gt; l = lambda i: a[i]</p>
<p>&gt;&gt;&gt; l</p>
<p>&lt;function at="" 0x100585a68=""&gt;</p>
<p>&gt;&gt;&gt; H = [(1, 2), (3, 4)]</p>
<p>&gt;&gt;&gt; [l(0) + l(1) for a in H]</p>
<p>Traceback (most recent call last):
  File "", line 1, in 
  File "", line 1, in 
  File "", line 1, in 
NameError: global name 'a' is not defined</p>
<p>&gt;&gt;&gt; list((l(0) + l(1) for a in H))</p>
<p>Traceback (most recent call last):
  File "", line 1, in 
  File "", line 1, in 
  File "", line 1, in 
NameError: global name 'a' is not defined</p>
<p>[/code]</p>
<p>So in Python 3, both the list comprehension and the generator comprehensions act the same, which is not too surprising.  I guess I should recode that piece of code to make it future proof, although this doesn't seem easy at the moment, and it may require converting a one-liner into a six-liner.  If you are interested, the piece of code is <a href="http://github.com/asmeurer/sympy/blob/15c3675ff67be854c12c349ed9034f12bb2f5247/sympy/integrals/risch.py#L297">here</a>.</p>
<p>So can anyone provide any insight into what is going on with that lambda function?  Running it with the <code>-3</code> switch to <code>python2.6</code> didn't give any warnings related to it.  </p>
<p><strong>Update:</strong> As I noted in a <a href="http://asmeurersympy.wordpress.com/2010/06/16/strange-python-behavior-can-someone-please-explain-to-me-what-is-going-on-here/#comment-121">comment</a>, I figured out how to make this future-proof.  I need to convert it from </p>
<p>[code language="py"]</p>
<p>def residue_reduce_derivation(H, D, x, t, z):
    lambdafunc = lambda i: i*derivation(a[1], D, x, t).as_basic().subs(z, i)/ \
         a[1].as_basic().subs(z, i)
    return S(sum([RootSum(a[0].as_poly(z), lambdafunc) for a in H]))
[/code]</p>
<p>to</p>
<p>[code language="py"]</p>
<p>def residue_reduce_derivation(H, D, x, t, z):
    return S(sum((RootSum(a[0].as_poly(z), lambda i: i*derivation(a[1], D, x, t).as_basic().subs(z, i)/ \
        a[1].as_basic().subs(z, i)) for a in H)))
[/code]</p>
<p>Thanks to all the commenters for the explanations.  </p>
<p>Also, you may have noticed that I discovered that if you use <tt>[code]</tt> instead of <tt>&lt;code&gt;</tt>, you get these nicer code blocks that <em>actually respect indentation!</em>  Now I just need to figure out how to make them syntax highlight Python code.</p>
<p><strong>Update 2:</strong> <tt>[code='py']</tt> colors it!  Sweet!</p>
<p><strong>Update 3:</strong> I just discovered that SymPy has a <code>Lambda()</code> object that handles this better.  In particular, it pretty prints the code, and is what is already being used for <code>RootSum()</code> in the rational function integrator, at least in Mateusz's polys9.  </p>
<p>[code language="py"]</p>
<p>&gt;&gt;&gt; integrate(1/(x**5 + 1), x)</p>
<p>log(1 + x)/5 + RootSum(625<em>_t<strong>4 + 125*_t</strong>3 + 25</em>_t<em><em>2 + 5</em>_t + 1, Lambda(_t, _t</em>log(x + 5*_t)))                                                   </p>
<p>[/code]</p>
<p>Still, this has been a very good learning experience.  </body></html></p>